File: 860.柠檬水找零.c
/*
 * @lc app=leetcode.cn id=860 lang=c
 *
 * [860] 柠檬水找零
 */

// @lc code=start

bool lemonadeChange(int *bills, int billsSize)
{
  int five, ten, twenty, i;
  five = ten = twenty = 0;
  for (i = 0; i < billsSize; i++)
  {
    if (bills[i] == 5)
      five++;
    else if (bills[i] == 10)
    {
      ten++;
      if (five == 0)
        return false;
      else
        five--;
    }
    else
    {
      twenty++;
      if (ten == 0)
      {
        if (five >= 3)
          five -= 3;
        else
          return false;
      }
      else
      {
        ten--;
        if (five == 0)
          return false;
        else
          five--;
      }
    }
  }
  return true;
}
// @lc code=end

File: store/860.柠檬水找零.js
/*
 * @lc app=leetcode.cn id=860 lang=javascript
 *
 * [860] 柠檬水找零
 */

// @lc code=start
/**
 * @param {number[]} bills
 * @return {boolean}
 */
var lemonadeChange = function (bills) {
  const money = [5, 10, 20];
  let map = new Map();
  money.forEach((item) => {
    map.set(item, 0);
  });
  for (let item of bills.values()) {
    const obj = {
      5: () => {
        map.set(item, map.get(item) + 1);
      },
      10: () => {
        if (map.get(5) > 0) map.set(5, map.get(5) - 1);
        else return false;
        map.set(item, map.get(item) + 1);
      },
      20: () => {
        if (map.get(5) >= 1 && map.get(10) >= 1) {
          map.set(5, map.get(5) - 1);
          map.set(10, map.get(10) - 1);
        } else if (map.get(5) >= 3) map.set(5, map.get(5) - 3);
        else return false;
        map.set(item, map.get(item) + 1);
      },
    };
    if (obj[item]() === false) return false;
  }
  return true;
};
// @lc code=end
